### PR TITLE
feat(v4-sdk): add hooks utility class

### PR DESCRIPTION
**Pull request by ewilz**
## PR Scope

Please title your PR according to the following types and scopes following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/):

- `fix(SDK name):` will trigger a patch version
- `chore(<type>):` will not trigger any release and should be used for internal repo changes
- `<type>(public):` will trigger a patch version for non-code changes (e.g. README changes)
- `feat(SDK name):` will trigger a minor version
- `feat(breaking):` will trigger a major version for a breaking change

## Description

_[Summary of the change, motivation, and context]_

## How Has This Been Tested?

_[e.g. Manually, E2E tests, unit tests, Storybook]_

## Are there any breaking changes?

_[e.g. Type definitions, API definitions]_

If there are breaking changes, please ensure you bump the major version Bump the major version (by using the title `feat(breaking): ...`), post a notice in #eng-sdks, and explicitly notify all Uniswap Labs consumers of the SDK.

## (Optional) Feedback Focus

_[Specific parts of this PR you'd like feedback on, or that reviewers should pay closer attention to]_

## (Optional) Follow Ups

_[Things that weren't addressed in this PR, ways you plan to build on this work, or other ways this work could be extended]_

### CODE REVIEW

File: sdks/v4-sdk/src/hook.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
+export type HookPermissions = {
```
Comment (ewilz) [inline sdks/v4-sdk/src/hook.ts:1]: should find a nice way to shorthand this type and enum

**Comment by ewilz** [comment]
pushing up for posterity...still needs some cleaning + address checks

File: sdks/v4-sdk/src/hook.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  [HookOptions.AfterAddLiquidity]: 10,
+  [HookOptions.BeforeAddLiquidity]: 11,
+  [HookOptions.AfterInitialize]: 12,
+  [HookOptions.BeforeInitialize]: 13,
+}
+
+export class Hook {
+  public static permissions(address: string): HookPermissions {
+    invariant(isAddress(address), 'invalid address')
+    if (/0x/.test(address)) address = address.slice(2)
```
Comment (jsy1218) [inline sdks/v4-sdk/src/hook.ts:43]: nit: i feel like some comments/explains would help make the code easier to follow, like why `address.slice(2)` (easier to follow) and `address.slice(36)` (slightly harder to follow)

**Review by jsy1218 — APPROVED** [review]

File: sdks/v4-sdk/src/hook.ts (hunk 2 above)
Comment (ewilz) [reply sdks/v4-sdk/src/hook.ts:43]: will do

File: sdks/v4-sdk/src/hook.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      afterRemoveLiquidityReturnsDelta: this._hasPermission(address, HookOptions.AfterRemoveLiquidityReturnsDelta),
+    }
+  }
+
+  public static hasPermission(address: string, hookOption: HookOptions) {
+    return this._hasPermission(this._normalizeHookAddress(address), hookOption)
+  }
+
+  private static _hasPermission(address: string, hookOption: HookOptions) {
+    // slice only the last 14 bits which are the hook flags and compare with the specific hookOption
```
Comment (hensha256) [inline sdks/v4-sdk/src/hook.ts:67]: nit but really its slicing the last 16 bits because hex characters are multiples of 4 right?

**Review by hensha256 — APPROVED** [review]

**Review by jsy1218 — APPROVED** [review]

**Pull request merged**